lib/mapping/MapEditUtils.cpp

```diff
@@ -0,0 +1,371 @@
+}
+
+int3 MapRect::topRight() const
+{
+	return int3(right(), y, z);
+}
+
+int3 MapRect::bottomLeft() const
+{
+	return int3(x, bottom(), z);
+}
+
+int3 MapRect::bottomRight() const
+{
+	return int3(right(), bottom(), z);
+}
+
+CTerrainSelection::CTerrainSelection(CMap* map) : CMapSelection(map)
+{
+
+}
+
+void CTerrainSelection::selectRange(const MapRect& rect)
+{
+	rect.forEach([this](const int3 pos)
+		{
+			this->select(pos);
+		});
+}
+
+void CTerrainSelection::deselectRange(const MapRect& rect)
+{
+	rect.forEach([this](const int3 pos)
+		{
+			this->deselect(pos);
+		});
+}
+
+void CTerrainSelection::setSelection(const std::vector<int3>& vec)
+{
+	for (auto pos : vec)
+		this->select(pos);
+}
+
+void CTerrainSelection::selectAll()
+{
+	selectRange(MapRect(int3(0, 0, 0), getMap()->width, getMap()->height));
+	selectRange(MapRect(int3(0, 0, 1), getMap()->width, getMap()->height));
+}
+
+void CTerrainSelection::clearSelection()
+{
+	deselectRange(MapRect(int3(0, 0, 0), getMap()->width, getMap()->height));
+	deselectRange(MapRect(int3(0, 0, 1), getMap()->width, getMap()->height));
+}
+
+CObjectSelection::CObjectSelection(CMap* map) : CMapSelection(map)
+{
+
+}
+
+const std::string TerrainViewPattern::FLIP_MODE_DIFF_IMAGES = "D";
+
+const std::string TerrainViewPattern::RULE_DIRT = "D";
+const std::string TerrainViewPattern::RULE_SAND = "S";
+const std::string TerrainViewPattern::RULE_TRANSITION = "T";
+const std::string TerrainViewPattern::RULE_NATIVE = "N";
+const std::string TerrainViewPattern::RULE_NATIVE_STRONG = "N!";
+const std::string TerrainViewPattern::RULE_ANY = "?";
+
+TerrainViewPattern::TerrainViewPattern() : diffImages(false), rotationTypesCount(0), minPoints(0)
+{
+	maxPoints = std::numeric_limits<int>::max();
+}
+
+TerrainViewPattern::WeightedRule::WeightedRule(std::string& Name) : points(0), name(Name)
+{
+	standardRule = (TerrainViewPattern::RULE_ANY == Name || TerrainViewPattern::RULE_DIRT == Name
+		|| TerrainViewPattern::RULE_NATIVE == Name || TerrainViewPattern::RULE_SAND == Name
+		|| TerrainViewPattern::RULE_TRANSITION == Name || TerrainViewPattern::RULE_NATIVE_STRONG == Name);
+	anyRule = (Name == TerrainViewPattern::RULE_ANY);
+	dirtRule = (Name == TerrainViewPattern::RULE_DIRT);
+	sandRule = (Name == TerrainViewPattern::RULE_SAND);
+	transitionRule = (Name == TerrainViewPattern::RULE_TRANSITION);
+	nativeStrongRule = (Name == TerrainViewPattern::RULE_NATIVE_STRONG);
+	nativeRule = (Name == TerrainViewPattern::RULE_NATIVE);
+}
+
+void TerrainViewPattern::WeightedRule::setNative()
+{
+	nativeRule = true;
+	standardRule = true;
+	//TODO: would look better as a bitfield
+	dirtRule = sandRule = transitionRule = nativeStrongRule = anyRule = false; //no idea what they mean, but look mutually exclusive
+}
+
+CTerrainViewPatternConfig::CTerrainViewPatternConfig()
+{
+	const JsonNode config(ResourceID("config/terrainViewPatterns.json"));
+	static const std::string patternTypes[] = { "terrainView", "terrainType" };
```

Comment:
use `std::array` instead

Reply:
This code is quite sensitive and also probably is being refactored by @DjWarmonger . Let not touch it at this point